core-server/src/main/java/org/glassfish/jersey/server/internal/scanning/PackageNamesScanner.java

```diff
@@ -107,15 +110,19 @@ public PackageNamesScanner(final ClassLoader classLoader, final String[] package
         this.classloader = classLoader;
 
         this.finderFactories = new HashMap<>();
-        add(new JarZipSchemeResourceFinderFactory());
-        add(new FileSchemeResourceFinderFactory());
-        add(new VfsSchemeResourceFinderFactory());
-        add(new BundleSchemeResourceFinderFactory());
 
-        // TODO - Services?
-        // for (UriSchemeResourceFinderFactory s : ServiceFinder.find(UriSchemeResourceFinderFactory.class)) {
-        //     add(s);
-        // }
+        ServiceFinder<UriSchemeResourceFinderFactory> servicesFound = ServiceFinder.find(UriSchemeResourceFinderFactory.class);
```

Comment:
This is dangerous. Suppose an environment with multiple JAX-RS applications. If one app adds a jar with UriSchemeResourceFinderFactory, it potentially breaks all other applications. 

I believe this should be allowed on an application level only. Fortunately, `PackageNamesScanner` is instantiated from the  `ResourceConfig`, so it would be possible to pass a boolean argument whether the feature is allowed or not. Something like:
`new PackageNamesScanner(packages, recursive, isProperty(ServerProperties.URI_SCHEME_RESOURCE_FINDER_FACTORY_SERVICE_ENABLED)
&& !isProperty(ServerProperties.METAINF_SERVICES_LOOKUP_DISABLE))`

Reply:
The dangerous thing here is that the PR implies the OR logic which excludes default resource finder factories when at least one custom factory is found. 
The documentation at the top of the PackageNameScanner states that default resource finders shall not be excluded, but only custom resource finder factories could be added if any are found. 

The whole algorithm of addition of such a factories works in the way that default factories come (are added) first, and then on top are added custom ones (if any). The add(..) method works so, that if new factory comes with existing scheme it will be used. So if custom factory comes with the scheme similar to the one in default implementation, the default will be replaced by a custom. 
Thus the whole PR shall bring the only change 

`for (UriSchemeResourceFinderFactory s : ServiceFinder.find(UriSchemeResourceFinderFactory.class)) {
    add(s);
 }`

it will be less dangerous at least.

And for sure tests shall be added for this modification.

Reply:
@senivam , so you're basically telling me that i cannot have any custom jar packaging?

Reply:
@driverpt nope, what I'm telling is when you are having your custom jar packaging it has priority over default. However default shall be included BEFORE your custom. Which is commented below the TODO (lines 125 - 127) in the original PackageNamedScanner . So, when you uncomment it (lines 125 -127), it will work as described in the documentation and as you are requiring - default jar resource factory finder will come first and afterwards it will be overwritten by your custom (when it's found by service finder).

Reply:
But that way the `JarZipSchemeResourceFinderFactory` will always run and cause an Exception to be Thrown

Reply:
The whole sequence of the PackageNameScanner looks like:

1. Initialize default resource finders
2. (not implemented, the aim of the current PR) initialize custom resource finders
3. run (create(...) method of) initialized resource finders. 

points 1 and 2 are related. Initialization of default and custom resource finders is mutually exclusive.

So yes, **JarZipSchemeResourceFinderFactory** will be initialized but it won't be run (it's create(...) method won't be called) if there would be found a custom resource finder which implements "**jar**", "**zip**", "**wsjar**" schemes. Because custom resource finder would remove default one from the **finderFactories** hash map.